DataModel/player.py

```diff
@@ -35,28 +37,95 @@ def __init__(self, name, position, team):
 
         self.can_ask_for_trump = False
 
+        self.earning_capacity = 0
+        self.trump_choice_card = None
+
         # Beliefs about Trump suit.
         # Beliefs about distribution of cards of other hands.
         self.belief = Belief(position.value)
 
-    # def suggest_move(self):
-    #     _valid_cards = self.game_state.valid_cards
-    #     _bid = self.game_state.bid
-    #     _metadata = self.game_state.metadata
-    #     _phase = _metadata.game_phase
-    #
-    #     if _phase == GamePhase.Bidding:
-    #         earning_capacity = self.belief.get_capacity(self.cards)
-    #         min_bid = _bid.minimumNextBid
-    #         if earning_capacity >= min_bid:
-    #             return earning_capacity
-    #
-    #     suggested_move = random.choice(_valid_cards)
-    #     return suggested_move
-
-    def rethink_belief(self, scenario, common_knowledge):
+    def suggest_move(self, game_state):
+        _game_state = GameState()    # weakref.ref(game_state)
+        _valid_cards = _game_state.valid_cards
+        _bid = _game_state.bid
+        _metadata = _game_state.metadata
+        _phase = _metadata.game_phase
+
+        if _phase == GamePhase.Bidding:
+            self.earning_capacity, self.trump_choice_card = belief.get_strength(self.cards)
+            if self.earning_capacity >= _bid.minimumNextBid:
+                return self.earning_capacity
+            else:
+                return -1
+
+        elif _phase == GamePhase.TrumpSelection:
+            _trump_card_id = self.trump_choice_card.id
+            assert _trump_card_id in _valid_cards
+            return _trump_card_id
+
+        elif _phase == GamePhase.Playing:
+            # TODO: Think and do something.
+            suggested_move = random.choice(_valid_cards)
+            return suggested_move
+
+        elif _phase == GamePhase.Over:
+            a = 5
+            # Throw exception??
+        # end if else clauses
+    # end suggest_move()
+
+    def rethink_belief(self, game_state, scenario, move, move_by):
+        _game_state = game_state  # weakref.ref(game_state) # GameState() #
+        _common_knowledge = _game_state.common_knowledge
+
         if scenario == "CardsDealt":
-            self.belief.my_hand_cards(self.cards, common_knowledge)
+            self.belief.my_hand_cards(self.cards, _common_knowledge)
         elif scenario == "CardPlayed":
-            a = 5
-            # something
+            # TODO : Looks like only truth is used to update beliefs. Try some rationality assumptions.
+            _current_player = move_by
+            _card_played_id = move.id   # _card_played_eng = move.eng ENG_TO_ID
+            self.belief.has_card(_card_played_id, _current_player)
+
+            # Check if there was a colour cut-off.
+            _carpet = _game_state.carpet
+            if move.suite != _carpet.suite:
+                self.belief.not_has_suite(_current_player, _carpet.suite)
+
+        elif scenario == "BidPlaced":
```

Comment:
Looks like a lot of if else in this piece. Whenever you get time, think if this can become better by using a different design.

Reply:
do you have any suggestions for a better design?